broker: ensure subtree restart upon loss of router node

garlick opened the pull request:
This adds code to handle the case described in #3608, where a broker might be forever cut off from the flux instance if its TBON parent restarts without informing it (for example a kill -9 or a node crash).  With this change, the parent detects the child trying to communicate without first executing the hello handshake to register its uuid, and sends a KEEPALIVE_DISCONNECT message to ensure it stops.  Systemd would then be able to restart it.

Also fixed a problem with the crashed broker properly rejoining _its_ parent, discovered during testing.

There is also some minor overlay cleanup here.

This is built on top of pr #3843.  Once that is merged, this can be rebased.

garlick commented:
I will work a bit on the testing here as there are some code paths in the diff that should definitely have coverage.

grondo commented:
> I will work a bit on the testing here as there are some code paths in the diff that should definitely have coverage.


Ok, that was going to be my one comment so I'll hold off reviewing for now.

grondo commented:
I'm getting reproducible failures from `t3306-system-routercrash` on my system.
Must be my environment because all tests appear to have passed in CI...

Any idea why the `--wait` option to `flux overlay status` doesn't appear to be working here?


```
 grondo@asp:~/git/f.git/t$ git describe
v0.28.0-211-g453cda11b
 grondo@asp:~/git/f.git/t$ ./t3306-system-routercrash.t -d -v
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/01-setup.sh
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/flux-sharness.sh
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/01-setup.sh
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/flux-sharness.sh
expecting success: 
	flux exec flux setattr log-stderr-mode local

ok 1 - tell each broker to log to stderr

expecting success: 
	echo "local://$(flux getattr rundir)/local-2" >uri2 &&
	test $(FLUX_URI=$(cat uri2) flux getattr rank) -eq 2

ok 2 - construct FLUX_URI for rank 2 (child of 1)

expecting success: 
	$startctl kill 1 9 &&
	test_expect_code 137 $startctl wait 1

flux-start: 1 (pid 16898) Killed
ok 3 - kill -9 broker 1

expecting success: 
	$startctl run 1

broker.err[0]: rank 1 uuid efeea11f-eb52-4753-852a-a6f806da9143 lost, to be superceded by new hello
broker.err[0]: hello from rank 1 uuid 8a6c9873-b258-471c-a332-6a749a4379c4 status partial
ok 4 - restart broker 1

expecting success: 
	FLUX_URI=$(cat uri2) test_must_fail flux ping --count=1 1

broker.err[2]: parent disconnect
broker.err[2]: state-machine.monitor: No route to host
flux-ping: 1!broker.ping: No route to host
ok 5 - ping broker 1 via broker 2

expecting success: 
	test_might_fail $startctl wait 2

flux-start: 2 (pid 16914) exited with rc=1
ok 6 - broker 2 was disconnected

expecting success: 
	$startctl run 2

ok 7 - restart broker 2

expecting success: 
	run_timeout 10 flux overlay status -vvv --pretty --color --wait full

flux-overlay: Upstream Flux broker is offline. Try again later.
not ok 8 - wait for rank 0 to report subtree status of full
#	
#		run_timeout 10 flux overlay status -vvv --pretty --color --wait full
#	

expecting success: 
	flux ping --count=1 2

flux-ping: 2!broker.ping: Resource temporarily unavailable
not ok 9 - ping broker 2 via broker 0
#	
#		flux ping --count=1 2
#	

expecting success: 
        flux mini run -n3 -N3 /bin/true

ok 10 - run a 3 node job

# failed 2 among 10 test(s)
1..10
broker.err[0]: rc2.0: /bin/bash -c sh ./t3306-system-routercrash.t  --verbose --debug --root=/home/grondo/git/f.git/t/trash-directory.t3306-system-routercrash Exited (rc=1) 4.1s
flux-start: 0 (pid 16882) exited with rc=1
```

grondo commented:
I can't understand what's going on here, just to make sure the `overlay.health` RPC was being generated as expected, I added `FLUX_HANDLE_TRACE` and got

```
expecting success: 
	FLUX_HANDLE_TRACE=1 flux overlay status -vvv --pretty --color --wait full

--------------------------------------
> 0.00000
> overlay.health
> flags=topic,payload,route userid=unknown rolemask=none nodeid=0 matchtag=1
> {"wait":"full"}
--------------------------------------
< 0.00038
< overlay.health
< flags=topic,payload,route userid=1000 rolemask=owner errnum=11 matchtag=1
< Upstream Flux broker is offline. Try again later.
flux-overlay: Upstream Flux broker is offline. Try again later.
not ok 8 - wait for rank 0 to report subtree status of full
#	
#		FLUX_HANDLE_TRACE=1 flux overlay status -vvv --pretty --color --wait full
#	
```

I don't understand how that RPC can get this particular error -- isn't it just an RPC to rank 0, which is clearly online?

Maybe I should try a fresh clone of this branch.

garlick commented:
> I don't understand how that RPC can get this particular error -- isn't it just an RPC to rank 0, which is clearly online?

Exactly what I was about to say.  I don't t_think_ it can happen on rank 0.  Print out `$FLUX_URI` maybe?  Did something earlier on have a side effect?

grondo commented:
Ah!

```
expecting success: 
	echo FLUX_URI=$FLUX_URI &&
	FLUX_HANDLE_TRACE=1 flux overlay status -vvv --pretty --color --wait full

FLUX_URI=local:///tmp/flux-system-xy88hs/local-2
--------------------------------------
> 0.00000
> overlay.health
> flags=topic,payload,route userid=unknown rolemask=none nodeid=0 matchtag=1
> {"wait":"full"}
--------------------------------------
< 0.00048
< overlay.health
< flags=topic,payload,route userid=1000 rolemask=owner errnum=11 matchtag=1
< Upstream Flux broker is offline. Try again later.
flux-overlay: Upstream Flux broker is offline. Try again later.

```

FLUX_URI is still set to rank 2..

grondo commented:
This fixes it
```diff
diff --git a/t/t3306-system-routercrash.t b/t/t3306-system-routercrash.t
index 98345b3f3..507deac6a 100755
--- a/t/t3306-system-routercrash.t
+++ b/t/t3306-system-routercrash.t
@@ -38,7 +38,9 @@ test_expect_success 'restart broker 1' '
 '
 
 test_expect_success 'ping broker 1 via broker 2' '
+       (
        FLUX_URI=$(cat uri2) test_must_fail flux ping --count=1 1
+       )
 '
 
 test_expect_success 'broker 2 was disconnected' '
```

Because of the way sharness uses `eval` on tests, I guess env vars set on command line still get exported to the global shell environment. Ugh.

garlick commented:
Thanks, I was finding that very puzzling.  I'll go ahead and incorporate that fix and repush.

garlick commented:
OK, I squashed that fix and one other one that I had added incrementally to the same test and forced a push.
Oh and I noticed the same kind of thing in `t3305-system-rpctrack-up.t`  so fixed it there too in another commit.

grondo commented:
I wonder if a `flux --uri=URI ...` option would be useful when the environment  should be avoided. `flux(1)` could modify the env before starting the subcommand or calling `flux_open(3)`

grondo reviewed (approved):
LGTM!

garlick commented:
Thanks - I'm going to set MWP.
